lib/model/query/entities.js

```diff
@@ -0,0 +1,144 @@
+      JOIN datasets ON datasets."projectId" = f."projectId"
+    left outer join entities on datasets."id" = entities."datasetId"
+    WHERE datasets."name" = ${datasetName} AND sd."id" = ${submissionDefId}
+    LIMIT 1),
+  entity AS (
+    SELECT entities."datasetId", entities."uuid", entities."id" from entities
+    WHERE entities."uuid" = ${entityUuid}
+  )
+  select dataset."id" as "datasetId", entity."id" as "entityId" from dataset
+  FULL OUTER JOIN entity
+    ON dataset."id" = entity."datasetId"`)
+    .then(({ datasetId, entityId }) =>
+      (datasetId !== null && entityId == null));
+
+// Entrypoint to where submissions (a specific version) become entities
+const processSubmissionDef = (submissionDefId) => ({ Datasets, Entities, Submissions }) =>
+  new Promise((resolve, reject) => {
+    Entities.getDefBySubmissionDefId(submissionDefId)
+      .then((def) => (def.isDefined()
+        ? reject(Problem.user.entityViolation({ reason: 'This submission was already used to create an entity.' }))
+        : Submissions.getDefById(submissionDefId).then((s) => s.get())
+          .then((submissionDef) => Datasets.getFieldsByFormDefId(submissionDef.formDefId)
+            .then((fields) =>
+              ((fields.length === 0)
+                ? resolve()
+                : Entity.fromSubmissionXml(submissionDef.xml, fields)
+                  .then((partial) =>
+                    Entities.canInsertEntity(partial.aux.dataset, submissionDefId, partial.uuid)
+                      .then((okToInsert) =>
+                        (okToInsert
+                          ? Entities.createNew(partial, submissionDef)
+                            .then((entity) => resolve(entity))
+                            .catch((err) => reject(Problem.user.entityViolation({ reason: 'Entity could not be created', err: err.problemDetails })))
+                          : reject(Problem.user.entityViolation({ reason: 'It is not possible to create an entity from this data.' })))))
+                  .catch((problem) => reject(problem)))))));
+  });
+
+
+////////////////////////////////////////////////////////////////////////////////
+// GETTING ENTITIES
+
+const getDefBySubmissionDefId = (submissionDefId) => ({ maybeOne }) =>
+  maybeOne(sql`select * from entity_defs where "submissionDefId" = ${submissionDefId} limit 1`)
+    .then(map(construct(Entity.Def)));
+
+const getByUuid = (uuid) => ({ maybeOne }) =>
+  maybeOne(sql`select * from entities where "uuid" = ${uuid} limit 1`)
+    .then(map(construct(Entity)));
+
+
+////////////////////////////////////////////////////////////////////////////////
+// SERVING ENTITIES
+
+const _exportUnjoiner = unjoiner(Entity, Entity.Def);
+
+const streamForExport = (datasetId) => ({ stream }) =>
+  stream(sql`
+select ${_exportUnjoiner.fields} from entity_defs
+inner join entities
+  on entities.id = entity_defs."entityId"
+where
+  entities."datasetId" = ${datasetId}
+  and entity_defs.current=true
+order by entities.id`)
```

Comment:
Noted at https://github.com/getodk/central-backend/pull/656#issuecomment-1308028806.

Reply:
turned this into its own issue (https://github.com/getodk/central-backend/issues/767) since we have several other order-related issues in the 2023.2 release.